### PR TITLE
Contract Data Ingestion

### PR DESCRIPTION
**Pull request by harshasomisetty**
querying smart contract data, mainly all existing pools and respective data
better support for url trade paths (reflected in trade modal) and basic pair trade statistics

### CODE REVIEW

File: src/pages/pools/index.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  const { wallet, publicKey } = useWallet();
+  const { connection } = useConnection();
+  const [pools, setPools] = useState<ProgramAccount<T>[]>([]);
+  const [custodies, setCustodies] = useState<Record<string, Object | Null[]>>(
+    {}
+  );
+
+  const stats = useDailyPriceStats();
+
+  useEffect(() => {
```
Comment (nramadas) [inline src/pages/pools/index.tsx:27]: stuff this like this should go into its own hook. something like `usePools`

File: src/pages/trade/[pair].tsx [hunk 2]
```diff
@@ -43,8 +43,16 @@ export default function Page() {
   const router = useRouter();
   const { pair } = router.query;
 
-  let token: ReturnType<typeof getToken> = Token.SOL;
-  let currency: ReturnType<typeof getComparisonCurrency> = "usd";
+  if (!pair) {
```
Comment (nramadas) [inline src/pages/trade/[pair].tsx:46]: why is this necessary? why can't we just default to a pair in the case where one isn't provided in the url?

File: src/pages/trade/[pair].tsx [hunk 3]
```diff
@@ -55,10 +63,12 @@ export default function Page() {
     }
   }
 
+  // TOOD eur is not supported through token type
+  // TODO figure out how to reconcile usdc and usd pairs
   return (
     <SidebarLayout className="pt-11">
       <div>
-        <TradeSidebar />
+        <TradeSidebar inputPayToken={token} outputPayToken={currency} />
```
Comment (nramadas) [inline src/pages/trade/[pair].tsx:71]: looking at the way GMX works, the input pay token  _isn't_ the same as the what shows up on the chart. The chart should reflect the token of the position you want to enter, ie Long SOL, Short SOL, etc, but what you pay can be independently selected with respect to the price chart. On GMX, if you select a different pay token, the chart stays the same. In our case, both the chart and the URL should stay the same

File: src/pages/trade/index.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import React, { useEffect } from "react";
+import Router from "next/router";
+
+const IndexPage = () => {
+  useEffect(() => {
+    const { pathname } = Router;
+    if (pathname == "/trade") {
+      Router.push("/trade/SOL-USD");
```
Comment (nramadas) [inline src/pages/trade/index.tsx:8]: why is this necessary? what was the bug with defaulting the tokens before?

File: src/hooks/useDailyPriceStats.ts [hunk 5]
```diff
@@ -22,6 +22,37 @@ export const TOKEN_LIST = [
...
+  switch (address) {
+    case "So11111111111111111111111111111111111111112":
+      return Token.SOL;
+    case "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU":
+      return Token.USDC;
+  }
+}
+
+// function that maps token string to token enum
+export function getTokenGivenString(token: string) {
```
Comment (nramadas) [inline src/hooks/useDailyPriceStats.ts:35]: it looks like this function is meant to be a typeguard of sorts to cast a string as a `Token`. In that case, it might be better to name this function `asToken`. And, you will also want to add a default case at the end of the switch statement that throws. Something like `default: throw new Error('not a valid token')`.

File: src/hooks/useDailyPriceStats.ts [hunk 6]
```diff
@@ -22,6 +22,37 @@ export const TOKEN_LIST = [
   Token.Bonk,
 ];
 
+export function getTokenGivenAddress(address: string) {
```
Comment (nramadas) [inline src/hooks/useDailyPriceStats.ts:25]: a better name for this might be `tokenAddressToToken`. Also, might want to consider a default case, otherwise we'll be dealing with `undefined`s being returned from this method. Or, if that is ok, still having an explicit default case and returning `null` would be preferable

**Review by nramadas — CHANGES_REQUESTED** [review]

File: src/pages/trade/index.tsx (hunk 4 above)
Comment (harshasomisetty) [reply src/pages/trade/index.tsx:8]: was using dydx and binance navigation system instead of gmx, i think its a cleaner trading ux

File: src/pages/pools/index.tsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+      fetchData();
+    }
+  }, [wallet, publicKey]);
+
+  if (pools.length === 0) {
+    return <p>Loading...</p>;
+  }
+
+  return (
+    <div>
```
[the remark this answers is not in the document]
Comment (harshasomisetty) [reply src/pages/pools/index.tsx:68]: yes adopt that format

File: src/pages/trade/[pair].tsx (hunk 2 above)
Comment (harshasomisetty) [reply src/pages/trade/[pair].tsx:46]: yes the default is now defaulting in the index page
